### ISSUE TITLE
Numeric values are not usable by numerical functions in SPSS data files

### ISSUE THREAD

**Issue by jaohar** [hadley/haven#210]
Numeric values are not able to be operated on with numerical functions (e.g. descriptives) on SPSS files outputted by write_sav. Integers work without issue. As a work around, the value type can be changed to string and back and then the numerical functions work as intended.

Please let me know if I am doing something incorrectly or if this is a known issue. If you would like me to produce an example dataset, let me know what structure would be ideal.

Thank you very much for all your work on haven.

**Comment by evanmiller** [comment]
Hi, please let us know what version of haven you are using, and provide the smallest example file that demonstrates the issue. Your issue may be related to #133. If you are running haven from CRAN please install the latest GitHub version:

```{R}
devtools::install_github("hadley/haven")
```

**Comment by jaohar** [comment]
Hello, the GitHub version of haven fixes this problem. Thank you. However now my R session crashes when I try to save my dataset as an SPSS file (for both R version 3.3.0 and Rstudio on Mac OS 10.11.6). I've isolated this down to a single factor/character value. I don't see any special characters here and don't understand what could be the problem.

Please try to reproduce this crash using the code below and the attached file. Please rename the attached file "abc.csv". Apologies for the content of the variable, the data is for public health research.

Thanks very much for your time.

library(haven)
abc <- read.csv("abc.csv")
write_sav(abc, "abc.sav")

[abc.csv.txt](https://github.com/hadley/haven/files/413031/abc.csv.txt)

**Comment by evanmiller** [comment]
@jaohar Thanks for the update. I am able to reproduce the crash using the file you provided. For my own reference the backtrace is:

```
Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff93e2df06 __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff95d944ec pthread_kill + 90
2   libsystem_c.dylib             	0x00007fff8a1986df abort + 129
3   libsystem_c.dylib             	0x00007fff8a198856 abort_report_np + 181
4   libsystem_c.dylib             	0x00007fff8a1bea0c __chk_fail + 48
5   libsystem_c.dylib             	0x00007fff8a1be9dc __chk_fail_overflow + 16
6   libsystem_c.dylib             	0x00007fff8a1bef22 __memcpy_chk + 37
7   haven.so                      	0x000000010ce8d896 sav_begin_data + 3238
8   haven.so                      	0x000000010ce7e305 readstat_begin_row + 165
9   haven.so                      	0x000000010cea3705 Writer::write() + 1237
10  haven.so                      	0x000000010cea3033 write_sav_(Rcpp::Vector<19, Rcpp::PreserveStorage>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 67
11  haven.so                      	0x000000010cea7b7f haven_write_sav_ + 175
12  libR.dylib                    	0x00000001036d0827 do_dotcall + 327 (dotcode.c:1251)
13  libR.dylib                    	0x00000001036fde4c Rf_eval + 988 (eval.c:713)
14  libR.dylib                    	0x00000001037520fe Rf_evalList + 110 (eval.c:2287)
15  libR.dylib                    	0x00000001036fdd44 Rf_eval + 724 (Rinlinedfuns.h:85)
16  libR.dylib                    	0x0000000103755da3 do_begin + 451 (eval.c:1807)
17  libR.dylib                    	0x00000001036fe07e Rf_eval + 1550 (eval.c:685)
18  libR.dylib                    	0x0000000103752b25 Rf_applyClosure + 1493 (eval.c:1134)
19  libR.dylib                    	0x00000001036fdf87 Rf_eval + 1303 (eval.c:732)
20  libR.dylib                    	0x0000000103755da3 do_begin + 451 (eval.c:1807)
21  libR.dylib                    	0x00000001036fe07e Rf_eval + 1550 (eval.c:685)
22  libR.dylib                    	0x0000000103752b25 Rf_applyClosure + 1493 (eval.c:1134)
23  libR.dylib                    	0x00000001036fdf87 Rf_eval + 1303 (eval.c:732)
24  libR.dylib                    	0x00000001037870cc R_ReplDLLdo1 + 460 (main.c:391)
25  org.R-project.R               	0x0000000103567127 run_REngineRmainloop + 295
26  org.R-project.R               	0x000000010355b4fa -[REngine runREPL] + 138
27  org.R-project.R               	0x000000010354a4ff main + 815
28  libdyld.dylib                 	0x00007fff8ba9d5ad start + 1
```

It is possibly related to the length of the string variable since >255 character support is new.

**Comment by evanmiller** [comment]
Sure enough it seems to start crashing right at the 255 character mark.

@hadley any tips on turning off optimization? It looks like the stack trace is losing information due to inlining.

**Comment by evanmiller** [comment]
Based on my poking around, it appears the crash isn't due to a long value per se, but rather due to a long value _label_. I am not sure what SPSS's limits are but I should be able to fix the crash at least.

**Comment by evanmiller** [comment]
Fixed on the ReadStat side in https://github.com/WizardMac/ReadStat/commit/0ce38f6142c508e29884480b522810e3e7eeddff

**Comment by evanmiller** [comment]
(According to the PSPP docs the limit is 120 characters. Right now I truncate the value labels longer than 120 characters, but it may be better to return an error instead.)

**Comment by jaohar** [comment]
Please continue with the same behaviour of truncating the value labels and return a warning rather than an error. Errors will be annoying for everyone dealing with lengthy text-based data stored as factors by default. Thanks again.

**Comment by evanmiller** [comment]
Should be fixed with #209.
